### DISCUSSION TITLE
Why is there no readonly ref record struct?

### DISCUSSION THREAD

**Discussion by hugener** [dotnet/csharplang#5431] [General] [answered]
I'm an functional programming beginner, but wonder why there is no readonly ref record struct, now that we got record structs in C# 10?

Wouldn't that make for a powerful tool for FP in C# avoiding the cost of copying large structs/allocating record classes when doing recursion?

Not sure if I'm missing something obvious here, but for now I'm trying readonly ref struct, although I loose the simplified declaration that records have. Or why doesn't structs have record style declaration as well?

**Comment by CyrusNajmabadi** [comment]
You don't need a `ref` struct for this. Just use a readonly record struct, and pass using `in`.

**Reply by hugener** [reply]
Great thanks!

**Reply by hypeartist** [reply]
> You don't need a `ref` struct for this`.

Oh, sure I do if I need to have a member of, say, ```ReadOnlySpan<>``` type

**Reply by CyrusNajmabadi** [reply]
you can use a `readonly ref struct` for that case.

**Comment by 333fred** [comment, marked as answer]
`record struct S` implies implementing `IEquatable<S>`. `ref struct`s cannot implement interfaces.

**Reply by AnorZaken** [reply]
(Small update: ref structs allow interfaces now, however IEquatable<T> cannot accept a ref struct as it's T parameter, so still blocked.)

**Comment by aelij** [comment]
@333fred can we revisit this now that interfaces are supported? It could be a simple alternative to #2584, e.g. for passing `Span`s into `String.Create()`.

**Reply by 333fred** [reply]
We could, though I don't know how this would subsume #2584. Seems like that would still be wanted.
